### PR TITLE
Use source list from the compiler to track invalidated files for hot reload.

### PR DESCRIPTION
**Pull request by aam**
This builds upon recently added frontend server functionality that returns list of sources used for recent compilation landed in https://github.com/flutter/flutter/pull/29627 and eliminates the needs to parse pubspec/package files, scan whole directories for changes.

### CODE REVIEW

File: examples/flutter_gallery/lib/gallery/home.dart [hunk 1]
```diff
@@ -14,7 +14,7 @@ import 'backdrop.dart';
 import 'demos.dart';
 
 const String _kGalleryAssetsPackage = 'flutter_gallery_assets';
-const Color _kFlutterBlue = Color(0xFF003D75);
+const Color _kFlutterBlue = Color(0xFFFF3D75);
```
Comment (jonahwilliams) [inline examples/flutter_gallery/lib/gallery/home.dart]: Nit: revert unrelated change

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: examples/flutter_gallery/lib/gallery/home.dart (hunk 1 above)
Comment (aam) [reply examples/flutter_gallery/lib/gallery/home.dart]: thanks for quick feedback! fixed.

**Comment by aam** [comment]
@jonahwilliams PTAL if you have a chance - I had changed how invalidation works as I was fixing bugs in initial version.

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 2]
```diff
@@ -941,104 +936,30 @@ class HotRunner extends ResidentRunner {
...
-    final File packagesFile = fs.file(_packagesPath);
-    if (packagesFile.existsSync()) {
-      final int newPackagesUpdateTime = packagesFile.statSync().modified.millisecondsSinceEpoch;
-      // Hot reloading with an updated package will often times kill a non-trivial
-      // appliction. This _might_ work, given certain application size and package
-      // constraints, so instead of exiting we print a warning so that the user has
-      // some hint on what went wrong.
-      if (newPackagesUpdateTime > _packagesUpdateTime) {
-        printError('Warning: updated dependencies detected. The Flutter application will require a restart to safely use new packages.');
+      final int updatedAt = fs
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/run_hot.dart]: Since we're only checking the files the frontend server used in the last compilation, it is possible that one or more files were deleted since then right?

In this case, I believe statSync will return null

**Comment by jonahwilliams** [comment]
How does the frontend_server handle files from a multiroot scheme (google3 and dev/integration_tests/codegen) ?

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 2 above)
Comment (aam) [reply packages/flutter_tools/lib/src/run_hot.dart]: Oh, right, thanks! Fixed.

**Comment by aam** [comment]
> How does the frontend_server handle files from a multiroot scheme (google3 and dev/integration_tests/codegen) ?

When reporting sources(in depfile or in list of sources used by compiler) frontend server will map them back to physical file system.

**Comment by jonahwilliams** [comment]
> When reporting sources(in depfile or in list of sources used by compiler) frontend server will map them back to physical file system.

Perfect, even less work for us to do in the tool then
